lib/tiniergltf/tiniergltf.hpp

```diff
@@ -1354,4 +1370,123 @@ struct GlTF {
 	}
 };
 
+// std::span is C++ 20, so we roll our own little struct here.
+template <typename T>
+struct Span {
+	T *ptr;
+	uint32_t len;
+	bool empty() const {
+		return len == 0;
+	}
+	T *end() const {
+		return ptr + len;
+	}
+	template <typename U>
+	Span<U> cast() const {
+		return {(U *) ptr, len};
+	}
+};
+
+static Json::Value readJson(Span<const char> span) {
+	Json::CharReaderBuilder builder;
+	const std::unique_ptr<Json::CharReader> reader(builder.newCharReader());
+	Json::Value json;
+	JSONCPP_STRING err;
+	if (!reader->parse(span.ptr, span.end(), &json, &err))
+		throw std::runtime_error("invalid JSON");
+	return json;
+}
+
+inline GlTF readGlb(const char *data, std::size_t len, const UriResolver &resolveUri = uriError) {
+	struct Chunk {
+		uint32_t type;
+		Span<const uint8_t> span;
+	};
+
+	struct Stream {
+		Span<const uint8_t> span;
+
+		bool eof() const {
+			return span.empty();
+		}
+
+		void advance(uint32_t n) {
+			span.len -= n;
+			span.ptr += n;
+		}
+
+		uint32_t readUint32() {
+			if (span.len < 4)
+				throw std::runtime_error("premature EOF");
+			uint32_t res = 0;
+			for (int i = 0; i < 4; ++i)
+				res += span.ptr[i] << (i * 8);
+			advance(4);
+			return res;
+		}
+
+		Chunk readChunk() {
+			const auto chunkLen = readUint32();
+			if (chunkLen % 4 != 0)
+				throw std::runtime_error("chunk length must be multiple of 4");
+			const auto chunkType = readUint32();
+
+			auto chunkPtr = span.ptr;
+			if (span.len < chunkLen)
+				throw std::runtime_error("premature EOF");
+			advance(chunkLen);
+			return {chunkType, {chunkPtr, chunkLen}};
+		}
+	};
+
+	constexpr uint32_t MAGIC_glTF = 0x46546C67;
+	constexpr uint32_t MAGIC_JSON = 0x4E4F534A;
+	constexpr uint32_t MAGIC_BIN = 0x004E4942;
```

Comment:
I would have spent less time looking at this code if it wasn't inside the function

Reply:
Well, I do really want to namespace these definitions to within the function; I don't want to expose them. The function scope fits this perfectly.